test/e2e/testsuites/testsuites.go

```diff
@@ -1023,21 +1023,22 @@ func (t *TestAzVolumeAttachment) Cleanup() {
 		framework.ExpectNoError(err)
 	}
 
-	time.Sleep(time.Duration(1) * time.Minute)
-	nodeNames, err := t.azclient.AzDriverNodes(t.namespace).List(context.Background(), metav1.ListOptions{})
-	framework.ExpectNoError(err)
-	for _, nodeName := range nodeNames.Items {
```

Comment:
There is no assurance that replica AzVolumeAttachment objects will attach only to the test AzDriverNodes created from `replicaNodeNames`. It can attach to any AzDriverNode in `azure-disk-csi` namespace. So this part should not be removed. Otherwise, it will not properly clean up

Reply:
This is the part that was actually causing the bug. Whenever we have multiple volume attachments the cleanup for the first az volume attachment deletes one of the attachments and all the az nodes and the cleanup for the other volume attachments was not entering this loop because all the nodes were already deleted. So similarly, this could happen to the replicas as well, so I don't think adding this loop back would solve the issue entirely. I'll look into the logic for `replicaNodeNames` to see if I can come up with something else, let me know if you have another approach in mind.

Reply:
I am not sure if I am following correctly. So, as it currently stands, no AzDriverNode is being deleted in the loop as the purpose of the loop is the remove all AzVolumeAttachments that is linked to a specified volume. And the AzDriverNodes in `replicaNodeNames` are deleted after this loop. Why are future cleanup calls not entering the loop? There still should be AzDriverNodes, either those generated from the node plugin, and those created from that specific test case where the clean up is called with new `TestAzVolumeAttachment` is set up. Could you elaborate?

Reply:
Since there is 1-1 mapping between `AzDriverNodes` and `nodes` in the cluster, not all tests create extra `AzDriverNodes` for each `AzVolumeAttachment`. This cleanup deletes all/some `AzDriverNodes` and thus consecutive calls to cleanup don't enter the loop as `AzDriverNodes` are deleted. I think the approach to clean up here needs to change. Maybe there should be separate setup and cleanup for `AzDriverNodes` and `AzVolumeAttachments`.

Reply:
Since node controller is already creating `AzDriverNodes` maybe `SetupWithAzVolumeAttachment` should not create any new instances of `AzDriverNodes` at all. If I understand correctly, this is not the expected behavior anyways.

Reply:
What I am trying to say is that it doesn't. https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/2b4608edb60936f5bba27c39d53d0029f79aac37/test/e2e/testsuites/testsuites.go#L1031
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/2b4608edb60936f5bba27c39d53d0029f79aac37/test/e2e/testsuites/testsuites.go#L1040
Only `AzVolumeAttachment` objects are deleted in this loop and AzDriverNodes with name in `replicaNames` are deleted. So there should never be a case where future clean up calls don't enter the loop unless some other operation deleted the AzDriverNodes generated upon node plugin registration.

Reply:
> Since node controller is already creating `AzDriverNodes` maybe `SetupWithAzVolumeAttachment` should not create any new instances of `AzDriverNodes` at all. If I understand correctly, this is not the expected behavior anyways.

You are right it is not an expected behavior, the only reason the test is currently framed this way is because AzVolumeAttachment controller's `Attach` and `Detach` are empty operations until controller provisioner PR is merged and it was a convenient way of assuring we have at least `n` number of AzDriverNodes to which replica attachments can be attached to. Once the PR is merged, these empty fake AzDriverNodes can no longer serve its purpose because there are no underlying physical nodes for these fake CRIs. Once the PR is merged, these fake AzDriverNodes will be deleted from the test and the number of replicas and the test will be replying on AzDriverNode CRIs created upon node plugin registration.

Reply:
Look at this case: 

node#1, node#2, node#3
azvolatt#1 (primarynode: node#1, replicas: node#2, node#3)
azvolatt#2 (primarynode: node#1, replicas: node#2, node#3) 
azvolatt#3 (primarynode: node#1, replicas: node#2, node#3) 

Call to delete one of the azvolatt#3 would delete all the nodes in the cluster, then when we go to delete the azvolatt#2 there will be no nodes in the cluster and it would not enter the loop to delete the azvolatt either because of this condition. 
 
And there can be the case when these nodes are the nodes generated by the node plugin because there really shouldn't be more nodes it the cluster than those generated by the node plugin. Every AzVolumeAttachment is not expected to have their dedicated AzNodeDriver. That's why I think the cleanup should not rely on the list of existing cluster nodes.

Reply:
Ok, I get what you are saying. I was confused because I didn't know you were referring to cases where there were multiple `TestAzVolumeAttachment`s, and this wasn't something I considered when I wrote the test cases. But you are right, in the above scenario, the current design would be problematic.

For your comment on 
```
And there can be the case when these nodes are the nodes generated by the node plugin because in the e2e tests there really shouldn't be more nodes it the cluster than those generated by the node plugin. Every AzVolumeAttachment is not expected to have their dedicated AzNodeDriver. That's why I think the cleanup should not rely on the list of existing cluster nodes.
```
please refer to my comment above.

Reply:
Since the scheduler extender tests also rely on this setup, I think it is more reasonable to get the number of cluster nodes (through ListNodeNames function) and skip the tests if itrequire more nodes then currently exist in the cluster rather then creating fake nodes. As of now, the controller tests require at most 3 nodes, so I don't expect that any tests would be skipped because this additional check and we can separate the cleanup of `AzDriverNodes` and `AzVolumeAttachments`.

Reply:
Let me know what you think!

Reply:
In that case, I don't think we even need a cleanup for `AzDriverNodes.` And all we would need to do, would be to create a single AzVolume and a single AzVolumeAttachment in each setup with primary role as its requested role. And the controller should be left to do the rest.

Reply:
I totally agree on the point that we should determine whether we run specific test cases based on the number of existing AzDriverNodes registered from node plugins

Reply:
Awesome! Yes, I think this approach will save us a lot of headache on cleanup :D